File: examples/python/basic_usage.py
"""
nulpoint Gateway — Python examples
====================================

Prerequisites:
    pip install openai

Run the gateway first:
    OPENAI_API_KEY=sk-... make run

Then run this file:
    python examples/python/basic_usage.py
"""

from openai import OpenAI

# Point the OpenAI client at the local gateway.
# The api_key is passed through to the provider but can be any non-empty string.
client = OpenAI(
    api_key="any-string",           # gateway handles auth with providers
    base_url="http://localhost:8080/v1",
)


def basic_chat() -> None:
    """Single-turn chat completion."""
    print("=== Basic chat ===")
    response = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": "What is the capital of France?"}],
    )
    print(response.choices[0].message.content)
    print(f"Tokens used: {response.usage.total_tokens}")
    print()


def multi_turn_chat() -> None:
    """Multi-turn conversation."""
    print("=== Multi-turn chat ===")
    messages = [
        {"role": "system", "content": "You are a helpful assistant who responds briefly."},
        {"role": "user",   "content": "My name is Alice."},
        {"role": "assistant", "content": "Nice to meet you, Alice!"},
        {"role": "user",   "content": "What's my name?"},
    ]
    response = client.chat.completions.create(model="gpt-4o", messages=messages)
    print(response.choices[0].message.content)
    print()


def streaming_chat() -> None:
    """Streaming response — tokens are printed as they arrive."""
    print("=== Streaming chat ===")
    stream = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": "Count slowly from 1 to 5."}],
        stream=True,
    )
    for chunk in stream:
        content = chunk.choices[0].delta.content
        if content:
            print(content, end="", flush=True)
    print("\n")


def use_anthropic_model() -> None:
    """Route to Anthropic Claude by specifying a Claude model name."""
    print("=== Anthropic Claude ===")
    response = client.chat.completions.create(
        model="claude-3-5-sonnet",
        messages=[{"role": "user", "content": "Hello! What model are you?"}],
        max_tokens=128,
    )
    print(response.choices[0].message.content)
    print()


def demonstrate_caching() -> None:
    """
    The gateway caches identical non-streaming requests.
    The second call returns instantly from cache (check X-Cache: HIT header).
    """
    print("=== Response caching ===")
    question = "What is the tallest mountain on Earth?"

    for i in range(2):
        response = client.chat.completions.create(
            model="gpt-4o",
            messages=[{"role": "user", "content": question}],
        )
        # The underlying HTTP response carries an X-Cache header:
        # first request → MISS, second → HIT.
        print(f"Request {i + 1}: {response.choices[0].message.content[:60]}...")
    print()


def embeddings() -> None:
    """
    Embed text using the /v1/embeddings endpoint.
    Supported models: text-embedding-3-small, text-embedding-3-large,
    text-embedding-ada-002 (OpenAI), mistral-embed, text-embedding-004 (Gemini).
    """
    print("=== Embeddings ===")

    # Single input — plain string
    response = client.embeddings.create(
        model="text-embedding-3-small",
        input="The quick brown fox jumps over the lazy dog.",
    )
    vec = response.data[0].embedding
    print(f"Single input → dims: {len(vec)}, first 4: {[round(v, 4) for v in vec[:4]]}")

    # Batch input — list of strings; one vector is returned per item
    response = client.embeddings.create(
        model="text-embedding-3-small",
        input=["Hello world", "How are you?", "Embeddings are useful."],
    )
    print(f"Batch input  → {len(response.data)} vectors, dims: {len(response.data[0].embedding)}")
    print(f"Tokens used: {response.usage.total_tokens}")
    print()


if __name__ == "__main__":
    basic_chat()
    multi_turn_chat()
    streaming_chat()
    use_anthropic_model()
    demonstrate_caching()
    embeddings()

File: examples/js/basic_usage.mjs
/**
 * nulpoint Gateway — JavaScript / Node.js examples
 * ===================================================
 *
 * Prerequisites:
 *   npm install openai
 *
 * Run the gateway first:
 *   OPENAI_API_KEY=sk-... make run
 *
 * Then run this file:
 *   node examples/js/basic_usage.mjs
 */

import OpenAI from "openai";

// Point the OpenAI client at the local gateway.
const client = new OpenAI({
  apiKey: "any-string", // gateway forwards auth to providers
  baseURL: "http://localhost:8080/v1",
});

// ── Basic chat completion ────────────────────────────────────────────────────
async function basicChat() {
  console.log("=== Basic chat ===");

  const response = await client.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "What is the capital of Japan?" }],
  });

  console.log(response.choices[0].message.content);
  console.log(`Tokens used: ${response.usage.total_tokens}\n`);
}

// ── Multi-turn conversation ──────────────────────────────────────────────────
async function multiTurnChat() {
  console.log("=== Multi-turn chat ===");

  const response = await client.chat.completions.create({
    model: "gpt-4o",
    messages: [
      { role: "system",    content: "You are a concise assistant." },
      { role: "user",      content: "My favourite colour is blue." },
      { role: "assistant", content: "Got it!" },
      { role: "user",      content: "What is my favourite colour?" },
    ],
  });

  console.log(response.choices[0].message.content, "\n");
}

// ── Streaming response ───────────────────────────────────────────────────────
async function streamingChat() {
  console.log("=== Streaming chat ===");

  const stream = await client.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "Count from 1 to 5, one number per line." }],
    stream: true,
  });

  for await (const chunk of stream) {
    const content = chunk.choices[0]?.delta?.content ?? "";
    process.stdout.write(content);
  }
  console.log("\n");
}

// ── Use Anthropic Claude ─────────────────────────────────────────────────────
async function useAnthropicModel() {
  console.log("=== Anthropic Claude ===");

  const response = await client.chat.completions.create({
    model: "claude-3-5-sonnet",
    messages: [{ role: "user", content: "What model are you running on?" }],
    max_tokens: 128,
  });

  console.log(response.choices[0].message.content, "\n");
}

// ── Failover demo ────────────────────────────────────────────────────────────
// If the primary provider fails, the gateway automatically retries the next
// one in the fallback chain (openai → anthropic → gemini → mistral).
// No code changes needed — just set multiple provider API keys.
async function failoverDemo() {
  console.log("=== Failover (transparent) ===");
  console.log(
    "The gateway automatically routes to the next available provider.\n" +
    "Enable multiple provider keys to see failover in action.\n"
  );
}

// ── Embeddings ────────────────────────────────────────────────────────────────
// Supported models: text-embedding-3-small, text-embedding-3-large,
// text-embedding-ada-002 (OpenAI), mistral-embed, text-embedding-004 (Gemini).
async function embeddings() {
  console.log("=== Embeddings ===");

  // Single input — plain string
  const single = await client.embeddings.create({
    model: "text-embedding-3-small",
    input: "The quick brown fox jumps over the lazy dog.",
  });
  const vec = single.data[0].embedding;
  const first4 = vec.slice(0, 4).map((v) => v.toFixed(4)).join(", ");
  console.log(`Single input → dims: ${vec.length}, first 4: ${first4}`);

  // Batch input — array of strings; one vector returned per item
  const batch = await client.embeddings.create({
    model: "text-embedding-3-small",
    input: ["Hello world", "How are you?", "Embeddings are useful."],
  });
  console.log(`Batch input  → ${batch.data.length} vectors, dims: ${batch.data[0].embedding.length}`);
  console.log(`Tokens used: ${batch.usage.total_tokens}\n`);
}

// ── Run all examples ─────────────────────────────────────────────────────────
try {
  await basicChat();
  await multiTurnChat();
  await streamingChat();
  await useAnthropicModel();
  await failoverDemo();
  await embeddings();
} catch (err) {
  console.error("Error:", err.message);
  process.exit(1);
}
